pkg/streamer/file_test.go

```diff
@@ -367,96 +378,83 @@ func (t *testUtilSuite) TestRelaySubDirUpdated(c *C) {
+		meta = Meta{BinLogName: relayFiles[2], BinLogPos: binlogPos, BinlogGTID: binlogGTID}
+		var buf bytes.Buffer
+		enc := toml.NewEncoder(&buf)
+		c.Assert(enc.Encode(meta), IsNil)
+		c.Assert(utils.WriteFileAtomic(path.Join(subDir, utils.MetaFilename), buf.Bytes(), 0o644), IsNil)
+		fi, err4 := os.Stat(relayPaths[1])
+		c.Assert(err4, IsNil)
+		curSize := fi.Size()
+		// write old binlog file
+		err5 := ioutil.WriteFile(relayPaths[1], data, 0o600)
+		c.Assert(err5, IsNil)
 		c.Assert(len(errCh), Equals, 0)
-		c.Assert(len(updatePathCh), Equals, 1)
+		c.Assert(failpoint.Enable("github.com/pingcap/dm/pkg/streamer/CMPAlwaysReturn0", `return(true)`), IsNil)
+		//nolint:errcheck
+		defer failpoint.Disable("github.com/pingcap/dm/pkg/streamer/CMPAlwaysReturn0")
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			relayLogUpdatedOrNewCreated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], curSize, updatePathCh, errCh)
+		}()
 		up := <-updatePathCh
 		c.Assert(up, Equals, relayPaths[1])
+		c.Assert(failpoint.Disable("github.com/pingcap/dm/pkg/streamer/CMPAlwaysReturn0"), IsNil)
 	}()
-
-	// wait watcher started, update the file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(relayPaths[1], data, 0o600)
-	c.Assert(err, IsNil)
-	wg.Wait()
-
-	// 5. file created
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[1], relayFiles[1], size, updatePathCh, errCh)
-		c.Assert(len(errCh), Equals, 0)
-		c.Assert(len(updatePathCh), Equals, 1)
-		up := <-updatePathCh
-		c.Assert(up, Equals, relayPaths[2])
-	}()
-
-	// wait watcher started, create a new file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(relayPaths[2], data, 0o600)
-	c.Assert(err, IsNil)
 	wg.Wait()
 
-	// 6.
-	// directory created will be ignored
-	// invalid binlog name will be ignored
-	// rename file will be ignore
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		relaySubDirUpdated(ctx, watcherInterval, subDir, relayPaths[2], relayFiles[2], size, updatePathCh, errCh)
-		c.Assert(len(errCh), Equals, 0)
-		c.Assert(len(updatePathCh), Equals, 1)
-		up := <-updatePathCh
-		c.Assert(up, Equals, relayPaths[2])
-	}()
-
-	// wait watcher started, create new directory
-	time.Sleep(2 * watcherInterval)
-	err = os.MkdirAll(filepath.Join(subDir, "new-directory"), 0o700)
-	c.Assert(err, IsNil)
-
-	// wait again, create an invalid binlog file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(filepath.Join(subDir, "invalid-binlog-filename"), data, 0o600)
-	c.Assert(err, IsNil)
-
-	// wait again, rename an older filename, seems watch can not handle `RENAME` operation well
-	// time.Sleep(2 * watcherInterval)
-	// err = os.Rename(relayPaths[0], relayPaths[3])
-	// c.Assert(err, IsNil)
-
-	// wait again, update the file
-	time.Sleep(2 * watcherInterval)
-	err = ioutil.WriteFile(relayPaths[2], data, 0o600)
-	c.Assert(err, IsNil)
-	wg.Wait()
+	// 5. context timeout (no new write)
+	// revert meta file to old file
```

Comment:
why bother to revert to old file? does not do this causing a problem?

Reply:
Because the latest binlog file is currently binlog.002
But test `#4` above deliberately wrote the error binlog.003 into the meta 

The current test is to check that the error was returned as expected when the ctx was cancelled, so need to change the binlog file name in the wrong meta back to binlog.0002